library/Booster/Pattern/Rewrite.hs

```diff
@@ -271,13 +274,35 @@ applyRule pat@Pattern{ceilConditions} rule = runRewriteRuleAppT $ do
     -- in isolation). Stop if false, abort rewrite if indeterminate.
     let ruleRequires =
             concatMap (splitBoolPredicates . coerce . substituteInTerm subst . coerce) rule.requires
-        failIfUnclear = RuleConditionUnclear rule
         notAppliedIfBottom = RewriteRuleAppT $ pure NotApplied
     unclearRequires <-
-        catMaybes <$> mapM (checkConstraint failIfUnclear notAppliedIfBottom) ruleRequires
-    unless (null unclearRequires) $
-        failRewrite $
-            head unclearRequires
+        catMaybes <$> mapM (checkConstraint id notAppliedIfBottom) ruleRequires
+
+    -- check unclear requires-clauses in the context of known constraints (prior)
+    let prior = pat.constraints
+    mbSolver <- lift $ RewriteT $ (.smtSolver) <$> ask
+
+    case mbSolver of
+        Just solver -> do
+            checkAllRequires <- lift $ SMT.checkPredicates solver prior mempty (Set.fromList unclearRequires)
+
+            case checkAllRequires of
+                Nothing ->
+                    -- unclear even with the prior
+                    failRewrite $
+                        RuleConditionUnclear rule . coerce $
+                            foldl1 AndTerm $
+                                map coerce unclearRequires
+                Just False ->
+                    -- requires is actually false given the prior
+                    RewriteRuleAppT $ pure NotApplied
+                Just True ->
+                    -- can proceed
+                    pure ()
+        Nothing ->
+            unless (null unclearRequires) $
+                failRewrite $
+                    RuleConditionUnclear rule (head unclearRequires)
 
     -- check ensures constraints (new) from rhs: stop and return `Trivial` if
     -- any are false, remove all that are trivially true, return the rest
```

Comment:
Below this comment, we should actually use the solver on the prior and the ensured constraints and possibly return `Trivial` if these are `Unsat` together. Left for a follow-up PR (this will change the `test-vacuous` results).